// authenticare/client/index.js
"use strict";

var _request = _interopRequireDefault(require("./request"));

var _endpoints = require("../endpoints");

var _auth = require("./auth");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

module.exports = {
  signIn: signIn,
  logOff: _auth.logOff,
  register: register,
  isAuthenticated: _auth.isAuthenticated,
  getDecodedToken: _auth.getDecodedToken,
  getEncodedToken: _auth.getEncodedToken,
  getAuthorizationHeader: getAuthorizationHeader
};

function register(newUser, options) {
  var baseUrl = options && options.baseUrl;
  var url = "".concat(baseUrl || '').concat(_endpoints.registerUrl);
  return (0, _request["default"])(url, newUser);
}

function signIn(user, options) {
  var baseUrl = options && options.baseUrl;
  var url = "".concat(baseUrl || '').concat(_endpoints.signInUrl);
  return (0, _request["default"])(url, user);
}

function getAuthorizationHeader() {
  var token = (0, _auth.getEncodedToken)();
  return {
    Authorization: token ? "Bearer ".concat(token) : null
  };
}
